REFACTORING_RECOMMENDATIONS.md

```diff
@@ -0,0 +1,1205 @@
+    }
+
+    /**
+     * Override to add additional initialization logic.
+     */
+    protected initializeAdditional?(): MaybePromise;
+
+    private assertSupportedModesWithRequiredTags(): void {
+        ModeUtils.assertSupportedModes(this.state.supportedModes);
+
+        // Validate all required tags are present
+        for (const requiredTag of this.requiredModeTags) {
+            const hasTag = this.state.supportedModes.some(({ modeTags }) =>
+                modeTags.some(({ value }) => value === requiredTag),
+            );
+
+            if (!hasTag) {
+                throw new ImplementationError(
+                    `Provided supportedModes must include mode tag ${requiredTag}`
+                );
+            }
+        }
+
+        // Allow subclasses to add additional validation
+        this.validateAdditionalModeTags?.();
+    }
+
+    #assertMode = (newMode: number): void => {
+        ModeUtils.assertMode(this.state.supportedModes, newMode);
+    };
+
+    override changeToMode({
+        newMode
+    }: ModeBase.ChangeToModeRequest): MaybePromise<ModeBase.ChangeToModeResponse> {
+        const result = ModeUtils.assertModeChange(
+            this.state.supportedModes,
+            this.state.currentMode,
+            newMode
+        );
+
+        if (result.status === ModeBase.ModeChangeStatus.Success) {
+            this.state.currentMode = newMode;
+        }
+
+        return result;
+    }
+}
+```
+
+**Refactor Servers**:
+
+```typescript
+// Before: /packages/node/src/behaviors/dishwasher-mode/DishwasherModeServer.ts
+export class DishwasherModeServer extends DishwasherModeBehavior {
+    override initialize(): MaybePromise {
+        this.#assertSupportedModes();
+        ModeUtils.assertMode(this.state.supportedModes, this.state.currentMode);
+        this.reactTo(this.events.currentMode$Changing, this.#assertMode);
+    }
+
+    #assertSupportedModes() {
+        ModeUtils.assertSupportedModes(this.state.supportedModes);
+        if (!this.state.supportedModes.some(({ modeTags }) =>
+            modeTags.some(({ value }) => value === DishwasherMode.ModeTag.Normal),
+        )) {
+            throw new Error("Provided supportedModes need to include at least Normal mode tag");
+        }
+    }
+
+    #assertMode(newMode: number) {
+        ModeUtils.assertMode(this.state.supportedModes, newMode);
+    }
+
+    override changeToMode({ newMode }: ModeBase.ChangeToModeRequest): MaybePromise<ModeBase.ChangeToModeResponse> {
+        const result = ModeUtils.assertModeChange(this.state.supportedModes, this.state.currentMode, newMode);
+        if (result.status === ModeBase.ModeChangeStatus.Success) {
+            this.state.currentMode = newMode;
+        }
+        return result;
+    }
+}
+
+// After:
+import { ModeServerBase } from "../utils/ModeServerBase.js";
+import { DishwasherMode } from "#types";
+
+export class DishwasherModeServer extends ModeServerBase<DishwasherModeBehavior> {
```

Comment:
This loses inheritance of *Behavior.  Should be handled with composition of utility class instead